### ISSUE TITLE
'zero complete fields are not ignored' in v1.4.4 inflate number rows with no meaningful data 

### ISSUE THREAD

**Issue by targetpharma** [IUREDCap/redcap-etl#26]
[Version 1.4.4](https://github.com/IUREDCap/redcap-etl/releases/tag/1.4.4) introduced a change:

> Changed processing of complete fields, so that an exported data row from REDCap will be included in a database table that includes that complete field if the value of the field is zero, even if no other fields for that table are set

For reference, the commit with this change is [here](https://github.com/IUREDCap/redcap-etl/commit/f00be55e9ae9b491ae38cb88237460676c12c9ff).

For better or worse, our organization does not use the 'Repeating Events' feature. So every instrument is assigned a discreet number of events. In some cases, this number can be quite large (i.e., > 100).

The REDCap API will output a row for every instrument + event pairing. And in the case where no data has been entered in the instrument, the complete field defaults to 0.

With this change, the downstream database now has a row for every instrument + event pairing, even if no data was entered.

It might just be a preference, but we prefer the pre-1.4.4 behavior.

Is there an easy way to make this a configurable option? Further, could that option be configurable from the UI of the external module?

**Comment by mullen2** [comment]
Thanks for the feedback. Looking into this.

**Comment by mullen2** [comment]
@targetpharma: What I am seeing in my test case is that zero is only returned for a form complete field if the form/instrument was saved. If a form is never saved, then null/blank is returned for the form complete field and no record is generated by REDCap-ETL. Is that not what you are seeing?

**Comment by mullen2** [comment]
This is the record status dashboard in REDCap for my test project. For record 1001, I saved the "First Visit" and "Second Visit" Weight forms (without entering any data) in REDCap. Only these 2 records show up in the weight table generated by REDCap-ETL.
 
![image](https://user-images.githubusercontent.com/25109733/92603066-de707580-f27c-11ea-83be-d9923776f8f6.png)

**Comment by targetpharma** [comment]
@mullen2, thanks for looking into this and thanks for the update. Let me look more into our project setup. There might be something we are doing behind the scenes that is causing the API (and thus REDCap-ETL) to return a 0 for the complete field of forms that have never been visited.

**Comment by targetpharma** [comment]
@mullen2, what version of REDCap are you on? We are on an older version right now, 8.10.6. So there could possibly have been an API change between our version and what you are testing.

When I directly query the redcap_data table, the complete fields do not exist for form + event combos that have never been saved. But the API returns a row with 0 nonetheless.

![image](https://user-images.githubusercontent.com/20141935/92605039-3dcf8500-f27f-11ea-97ba-db5dc435c76f.png)

![image](https://user-images.githubusercontent.com/20141935/92605161-65bee880-f27f-11ea-8ce9-3e0b4e387d19.png)

Output from the API playground for the abstraction_tracking form:

```
abs_timepoint,abs_pages,abstractor_assigned,claim_date,abstraction_status,complete_date,abs_comments,saa_file,final_file,encrypt_key,abstraction_tracking_complete
0,55,jane_doe,,COMPLETE,2018-02-09,,[document],,xxx,1
3,19,john_doe,,COMPLETE,2018-05-11,,,,,1
,,,,,,,,,,0
,,,,,,,,,,0
,,,,,,,,,,0
,,,,,,,,,,0
,,,,,,,,,,0
,,,,,,,,,,0
,,,,,,,,,,0
,,,,,,,,,,0
,,,,,,,,,,0
,,,,,,,,,,0
,,,,,,,,,,0
,,,,,,,,,,0
,,,,,,,,,,0
,,,,,,,,,,0
,,,,,,,,,,0
,,,,,,,,,,0
,,,,,,,,,,0
,,,,,,,,,,0
,,,,,,,,,,0
,,,,,,,,,,0
,,,,,,,,,,0
,,,,,,,,,,
```

**Comment by mullen2** [comment]
@targetpharma Thanks for the additional information. I used REDCap 9.2.58, so it appears there may well be a difference between the versions. I am looking into adding an option to ignore empty, incomplete forms.

**Comment by mullen2** [comment]
@targetpharma I created a VM with REDCap 8.10.6, and I am still seeing the same behavior that I saw in REDCap 9.2.58 (i.e., no records for the form are stored in the database by REDCap-ETL if the form was not saved).

There may be some important difference between our projects, but I am not able to see your project. Would it be possible for you to create a simple test project (with fake data) that tests this issue? If the issue does not occur, perhaps you could see the difference between your test project and actual project that is causing the issue. If it does occur, then if you could post a REDCap XML dump of the test project, I could try it on my system.

**Comment by targetpharma** [comment]
Thanks @mullen2 for the due diligence. I'll try to put together some test materials on my end to reproduce in a simple fashion.

**Comment by mullen2** [comment]
@targetpharma I think I was able to recreate the issue. In my original test, I only had 1 form for each event. When I added a second form for the visit events, I noticed that if the second form for the event was saved, then the first form would be returned with a zero (instead of blank) form complete field value, even though the first form was never saved.

**Comment by mullen2** [comment]
@targetpharma FYI, a property to handle this issue (ignore_empty_incomplete_forms) has been added to the REDCap-ETL application, and I am working on adding this to the REDCap-ETL external module.

**Comment by targetpharma** [comment]
Thanks for the update @mullen2

**Comment by mullen2** [comment]
@targetpharma The external module update for this was just released (version 1.3.1). Thank you for your continuing, very helpful feedback!
